File: Solution Bellman-Ford Search.java

import java.util.Arrays;

public class Solution {

    private static final int NO_SUCH_ROUTE = -1;

    public int findCheapestPrice(int numberOfCities, int[][] flights, int start, int goal, int maxNumberOfStops) {
        return bellmanFordSearchForCheapestPriceWithMaxStops(numberOfCities, flights, start, goal, maxNumberOfStops);
    }

    private int bellmanFordSearchForCheapestPriceWithMaxStops(int numberOfCities, int[][] flights, int start, int goal, int maxNumberOfStops) {
        int[] minDistanceFromStart = new int[numberOfCities];
        Arrays.fill(minDistanceFromStart, Integer.MAX_VALUE);
        minDistanceFromStart[start] = 0;

        for (int i = 0; i <= maxNumberOfStops; ++i) {
            int[] temp = Arrays.copyOf(minDistanceFromStart, numberOfCities);

            for (int[] flight : flights) {
                int from = flight[0];
                int to = flight[1];
                int price = flight[2];

                if (minDistanceFromStart[from] != Integer.MAX_VALUE) {
                    temp[to] = Math.min(temp[to], minDistanceFromStart[from] + price);
                }
            }
            minDistanceFromStart = temp;
        }
        return minDistanceFromStart[goal] != Integer.MAX_VALUE ? minDistanceFromStart[goal] : NO_SUCH_ROUTE;
    }
}

File: Solution Bellman-Ford Search.js

/**
 * @param {number} numberOfCities
 * @param {number[][]} flights
 * @param {number} start
 * @param {number} goal
 * @param {number} maxNumberOfStops
 * @return {number}
 */
var findCheapestPrice = function (numberOfCities, flights, start, goal, maxNumberOfStops) {
    this.NO_SUCH_ROUTE = -1;
    return bellmanFordSearchForCheapestPriceWithMaxStops(numberOfCities, flights, start, goal, maxNumberOfStops);
};

/**
 * @param {number} numberOfCities
 * @param {number[][]} flights
 * @param {number} start
 * @param {number} goal
 * @param {number} maxNumberOfStops
 * @return {number}
 */
function bellmanFordSearchForCheapestPriceWithMaxStops(numberOfCities, flights, start, goal, maxNumberOfStops) {
    let minDistanceFromStart = new Array(numberOfCities).fill(Number.MAX_SAFE_INTEGER);
    minDistanceFromStart[start] = 0;

    for (let i = 0; i <= maxNumberOfStops; ++i) {
        const temp = Array.from(minDistanceFromStart);

        for (let [from, to, price] of flights) {
            if (minDistanceFromStart[from] !== Number.MAX_SAFE_INTEGER) {
                temp[to] = Math.min(temp[to], minDistanceFromStart[from] + price);
            }
        }
        minDistanceFromStart = temp;
    }
    return minDistanceFromStart[goal] !== Number.MAX_SAFE_INTEGER ? minDistanceFromStart[goal] : this.NO_SUCH_ROUTE;
}
